

## Memory load and processing time

The maximum memory load in this run was $`r (gc())[1,6] + (gc())[2,6]  `$ Mb .
It took `r format(Sys.time() - start.Time)` of processing time.


## Methods draft

For processing and quality control of the raw methylation data, a customized version of 
the CPACOR pipeline < PMID: 25853392> was used for quality control, 
data normalization and calculation of beta values, 
calculating principal components of the control probes for adjustment 
and exclusion of outliers based on the Inter-Quartile-Range. 
The threshold for the sample call rate was set to CHANGE. 
White blood cell type (WBC) sub-populations were estimated based on 100 CpG sites 
by the Houseman method REFERENCE as implemented in the minfi R package REFERENCE. 
XXXXX samples discordant for reported and genetic sex, based on CpGs on the X- and Y-chromosome,
was excluded from analyses. Additionally, quality control based on principal component analyses
of the control probes was conducted to detect samples with measurement failures.


## Credits

The code for the CPACOR analysis pipeline was adapted from Lehne et. al. (Genome Biology, 2015)
which was developed and written by Benjamin Lehne (Imperial College London) and Alexander Drong
(Oxford University).
The code for the low-level quality control was developed and written by Alexander Teumer
(University Medicine Greifswald/ Erasmus MC Rotterdam).
The code was combined into the current pipeline by Pascal Schlosser and Franziska Grundner-Culemann.
It is available at https://github.com/genepi-freiburg/Infinium-preprocessing.

See *A coherent approach for analysis of the Illumina HumanMethylation450 Bead Chip improves quality and performance in epigenome-wide association studies* by Lehne et. al., Genome Biology (2015)
for the basic idea. The method was then extended to EPIC arrays. 
Please cite this article in your publication. 
