## PART 2: Low-Level Quality Control


The quality control consists of two parts: 

1. The second part is based on control probes. Details are given in the [ILMN HD methylation assay protocol guide (15019519)](https://support.illumina.com/downloads/infinium_hd_methylation_assay_protocol_guide_(15019519_b).html).
2. The data is checked for sex mismatch. 



* By conducting a PCA on the control-probe information we obtain controlprobe scores. 

```{r}
#PCA of control-probe intensities
pca <- prcomp(na.omit(ctrl.all))
ctrlprobes.scores = pca$x
colnames(ctrlprobes.scores) = paste(colnames(ctrlprobes.scores), '_cp', sep='')

# ### extended savings 
 control.info <- data.frame(control)
```

* Then we look at the controls.
```{r}
#control probe based QC
# Details are given in the ILMN HD methylation assay protocol guide (15019519)

# bisulfate conversion: a non-CpG sequence with only one C at the target site will be converter to T by BS conversion. Therefore, converted control probes should be expressed (for each red/green channel) and unconverted not.
# BS-I
cprobes <- control.info[control.info$Type=="BISULFITE CONVERSION I" & (control.info$Color %in% c("Green","LimeGreen","Lime")),"Address"] # select converted green
controlQC <- data.frame(Cgreen=colMeans(ctrl.complete.Green.all[cprobes,]))
cprobes <- control.info[control.info$Type=="BISULFITE CONVERSION I" & (control.info$Color %in% c("Purple","Red","Tomato")),"Address"] # select converted red
controlQC <- cbind(controlQC,data.frame(Cred=colMeans(ctrl.complete.Red.all[cprobes,])))
cprobes <- control.info[control.info$Type=="BISULFITE CONVERSION I" & (control.info$Color %in% c("Blue","SkyBlue","Cyan")),"Address"] # select unconverted green
controlQC <- cbind(controlQC,data.frame(Ugreen=colMeans(ctrl.complete.Green.all[cprobes,])))
cprobes <- control.info[control.info$Type=="BISULFITE CONVERSION I" & (control.info$Color %in% c("Orange","Gold","Yellow")),"Address"] # select unconverted red
controlQC <- cbind(controlQC,data.frame(Ured=colMeans(ctrl.complete.Red.all[cprobes,])))

# BS-II (red intensity, green should be background level)
cprobes <- control.info[control.info$Type=="BISULFITE CONVERSION II","Address"] # select converted red
controlQC <- cbind(controlQC,data.frame(BSIIgreen=colMeans(ctrl.complete.Green.all[cprobes,])))
controlQC <- cbind(controlQC,data.frame(BSIIred=colMeans(ctrl.complete.Red.all[cprobes,])))

# hybridization + target removal: green channel only
cprobes <- control.info[grep("Hyb (High)",control.info$ExtendedType,value=F,fixed=T),"Address"] # select hyb high
controlQC <- cbind(controlQC,data.frame(HybH=ctrl.complete.Green.all[cprobes,]))
cprobes <- control.info[grep("Hyb (Low)",control.info$ExtendedType,value=F,fixed=T),"Address"] # select hyb low
controlQC <- cbind(controlQC,data.frame(HybL=ctrl.complete.Green.all[cprobes,]))
cprobes <- control.info[grep("Target Removal",control.info$ExtendedType,value=F,fixed=T),"Address"] # select target removals
controlQC <- cbind(controlQC,data.frame(TR=colMeans(ctrl.complete.Green.all[cprobes,])))

# specificity I - Limit PM and MM probes to specific channel (1-3 for green, 4-6 for red)
cprobes <- control_info[grep("GT Mismatch [4-6] \\(PM\\)",control_info$ExtendedType,value=F),"Address"] # select spec I signal
controlQC <- cbind(controlQC,data.frame(SpecIPMred=colMeans(ctrl_complete.Red.all[cprobes,]))) # specific match (red)
cprobes <- control_info[grep("GT Mismatch [1-3] \\(PM\\)",control_info$ExtendedType,value=F),"Address"] # select spec I signal
controlQC <- cbind(controlQC,data.frame(SpecIPMgreen=colMeans(ctrl_complete.Green.all[cprobes,]))) # specific match (green)

cprobes <- control_info[grep("GT Mismatch [4-6] \\(MM\\)",control_info$ExtendedType,value=F),"Address"] # select spec I signal
controlQC <- cbind(controlQC,data.frame(SpecIMMred=colMeans(ctrl_complete.Red.all[cprobes,]))) # specific match bg (red)
cprobes <- control_info[grep("GT Mismatch [1-3] \\(MM\\)",control_info$ExtendedType,value=F),"Address"] # select spec I signal
controlQC <- cbind(controlQC,data.frame(SpecIMMgreen=colMeans(ctrl_complete.Green.all[cprobes,]))) # specific match bg (green)

# specificity II
cprobes <- control.info[grep("Specificity ",control.info$ExtendedType,value=F),"Address"] # select spec II
controlQC <- cbind(controlQC,data.frame(SpecIIspec=colMeans(ctrl.complete.Red.all[cprobes,]))) # specific match (red)
controlQC <- cbind(controlQC,data.frame(SpecIIunspec=colMeans(ctrl.complete.Green.all[cprobes,]))) # unspecific match (green)

# extension
cprobes <- control.info[grep("Extension \\([CG]\\)",control.info$ExtendedType,value=F),"Address"] # select C,G (green)
controlQC <- cbind(controlQC,data.frame(ExtCG=colMeans(ctrl.complete.Green.all[cprobes,])))
cprobes <- control.info[grep("Extension \\([AT]\\)",control.info$ExtendedType,value=F),"Address"] # select A,T (red)
controlQC <- cbind(controlQC,data.frame(ExtAT=colMeans(ctrl.complete.Red.all[cprobes,])))

# staining red (DNP)/green (Biotin)
cprobes <- control.info[grep("DNP (High)",control.info$ExtendedType,value=F,fixed=T),"Address"] # select staining red
controlQC <- cbind(controlQC,data.frame(StainingRedH=ctrl.complete.Red.all[cprobes,]))
cprobes <- control.info[grep("Biotin (High)",control.info$ExtendedType,value=F,fixed=T),"Address"] # select staining green
controlQC <- cbind(controlQC,data.frame(StainingGreenH=ctrl.complete.Green.all[cprobes,]))
cprobes <- control.info[grep("DNP (Bkg)",control.info$ExtendedType,value=F,fixed=T),"Address"] # select staining red background
controlQC <- cbind(controlQC,data.frame(StainingRedB=ctrl.complete.Red.all[cprobes,]))
cprobes <- control.info[grep("Biotin (Bkg)",control.info$ExtendedType,value=F,fixed=T),"Address"] # select staining green background
controlQC <- cbind(controlQC,data.frame(StainingGreenB=ctrl.complete.Green.all[cprobes,]))

```

The first 3 rows of control probes information from QC contain the following information:

```{r}
controlQC[1:3,]
```

In the following control probes are checked. For a more detailed descriptin see e.g. the [ILMN HD methylation assay protocol guide (15019519)](https://support.illumina.com/downloads/infinium_hd_methylation_assay_protocol_guide_(15019519_b).html) or the [Illumina BeadArray Controls Reporter Software Guide ](https://support.illumina.com/downloads/beadarray-controls-reporter-software-guide-1000000004009.html),pages 6-8. Probes are evaluated by MA plots. BS-I and BS-II control probes check the DNA bisulfite conversion step.

```{r}
labels <- rownames(controlQC)
removeOutlier <- function(val1,val2,labels=NULL,xname="val1",yname="val2",title="",filterYout=1, SDrange=4 ,nullOffs=0.0001 ) {
    # SDrange is  range to define outliers
    # nullOffs is offset to avoid division by zero, log 0, etc.

    x <- log2(sqrt(val1*val2)+nullOffs)
    y <- log2((val1+nullOffs)/(val2+nullOffs))
    x.w <- x[which(x>log2(nullOffs))] # remove extreme values that are zero, i.e. for staining
    y.w <- y[which(x>log2(nullOffs))] # remove extreme values that are zero, i.e. for staining, remove them also from y (based on x-extremes)
    x.w <- Winsorize(x.w)
    y.w <- Winsorize(y.w)
    out <- which(x>mean(x.w)+sd(x.w)*SDrange) # upper X
    out <- c(out,which(x<mean(x.w)-sd(x.w)*SDrange)) # lower X
    if (filterYout==2) { out <- c(out,which(y>mean(y.w)+sd(y.w)*SDrange)) } # upper Y 
    out <- unique(c(out,which(y<mean(y.w)-sd(y.w)*SDrange))) # lower Y

    plot(x,y,xlab=bquote(log[2](sqrt(.(xname)%*%.(yname)))),ylab=bquote(log[2](.(xname)/.(yname))),main=title,xlim=c(min(x),max(x)+7))

    points(x[out],y[out],col="red")
    if (length(out)>0) {
        text(x[out],y[out],label=labels[out],col="red",adj=c(0,0),cex=0.7)
    }

    ret <- data.frame(sample=labels[out],x=x[out],y=y[out])
    return(ret)
}

if(filterOutlierCtrlQC){
# BS I-C
bufout <- removeOutlier(controlQC$Cred,controlQC$Ured,labels,"Cred","Ured","BS-I C/U (red)")
if (nrow(bufout)>0) samples.filtered <- rbind(samples.filtered,data.frame(bufout,filter="BS I-C (red)"))

# BS I-U
bufout <- removeOutlier(controlQC$Cgreen,controlQC$Ugreen,labels,"Cgreen","Ugreen","BS-I C/U (green)")
if (nrow(bufout)>0) samples.filtered <- rbind(samples.filtered,data.frame(bufout,filter="BS I-C (green)"))

# BS II
bufout <- removeOutlier(controlQC$BSIIred,controlQC$BSIIgreen,labels,"BSIIred","BSIIgreen","BS-II (red vs. green)")
if (nrow(bufout)>0) samples.filtered <- rbind(samples.filtered,data.frame(bufout,filter="BS II"))
}
```

We also check the Hybridisation of the amplified DNA to the array:

```{r}
if(filterOutlierCtrlQC){
# Hyb + TR
bufout <- removeOutlier(controlQC$HybH,controlQC$TR,labels,"HybH","TargetRemoval","Hybridization (green)")
if (nrow(bufout)>0) samples.filtered <- rbind(samples.filtered,data.frame(bufout,filter="Hyb_TR"))

# SpecI red
bufout <- removeOutlier(controlQC$SpecIPMred,controlQC$SpecIMMred,labels,"specific","unspecific","Specificity I (red)")
if (nrow(bufout)>0) samples.filtered <- rbind(samples.filtered,data.frame(bufout,filter="Specificity I (red)"))

# SpecI green
bufout <- removeOutlier(controlQC$SpecIPMgreen,controlQC$SpecIMMgreen,labels,"specific","unspecific","Specificity I (green)")
if (nrow(bufout)>0) samples.filtered <- rbind(samples.filtered,data.frame(bufout,filter="Specificity I (green)"))

# SpecII
bufout <- removeOutlier(controlQC$SpecIIspec,controlQC$SpecIIunspec,labels,"specific","unspecific","Specificity II")
if (nrow(bufout)>0) samples.filtered <- rbind(samples.filtered,data.frame(bufout,filter="Specificity II"))

# Extension
bufout <- removeOutlier(controlQC$ExtCG,controlQC$ExtAT,labels,"ExtCG","ExtAT","Extensions",2) # both on upper and lower Y outlier
if (nrow(bufout)>0) samples.filtered <- rbind(samples.filtered,data.frame(bufout,filter="Extensions"))

nullOffs=0.0001 # use the same value as defined in removeOutlier() function!
# Staining Red (do not filter upper left cluster when checking csv file)
bufout <- removeOutlier(controlQC$StainingRedH,controlQC$StainingRedB,labels,"RedH","RedB","Staining (red) - Upper left cluster is OK!")
bufout <- bufout[which(bufout$x>log2(nullOffs) | bufout$y<=0),] # do not remove upper left cluster
if (nrow(bufout)>0) samples.filtered <- rbind(samples.filtered,data.frame(bufout,filter="Staining Red (cross-check with Extension outliers!)"))

# Staining Green (do not filter upper left cluster when checking csv file)
bufout <- removeOutlier(controlQC$StainingGreenH,controlQC$StainingGreenB,labels,"GreenH","GreenB","Staining (green) - Upper left cluster is OK!")
bufout <- bufout[which(bufout$x>log2(nullOffs) | bufout$y<=0),] # do not remove upper left cluster
if (nrow(bufout)>0) samples.filtered <- rbind(samples.filtered,data.frame(bufout,filter="Staining Green (cross-check with Extension outliers!)"))

# Staining Red vs. Green
bufout <- removeOutlier(controlQC$StainingRedH,controlQC$StainingGreenH,labels,"RedH","GreenH","Staining (red vs. green)",2) # both on upper and lower Y outlier
if (nrow(bufout)>0) samples.filtered <- rbind(samples.filtered,data.frame(bufout,filter="Staining Green_Red (cross-check with Extension outliers!)"))

plot(1,1,type="n",xaxt="n",yaxt="n",xlab="",ylab="",bty="n")
text(1,1,label="cross-check with Extension outliers!")
# par(mfrow=c(1,1))

if (nrow(samples.filtered)>0) {
    samples.filtered <- merge(samples.filtered,datsamples,all.x=T,by.x="sample",by.y="Sentrix_Code")
    samples.filtered <- merge(samples.filtered,data.frame(callrate=sample.call),all.x=T,by.x="sample",by.y=0)    
    samples.filtered <- samples.filtered[order(samples.filtered$sample,samples.filtered$filter),]
}
write.csv(samples.filtered,file = paste0(outputdir,"/samples-filtered","_",projectname,"-",formatted.time,".csv"),row.names=F)

}
```


The following table lists the detected outliers identified by the quality control and can be found 
in the file 
\newline 
`r paste0(outputdir,"/samples-filtered.csv")`
\newline
 in a slightly expanded version. 

```{r}
if(nrow(samples.filtered) > 0){
	# remove redundant information
	posID = match("Sentrix_ID", colnames(samples.filtered))
	posPos = match("Sentrix_Position", colnames(samples.filtered))
	reduced.samples.filtered = samples.filtered[,-c(posID, posPos)]
	# find most interesting columns
	posName = match("Sample_Name", colnames(reduced.samples.filtered))
	posFilter = match("filter", colnames(reduced.samples.filtered))
	# print
	print(reduced.samples.filtered[,c(posName,posFilter)])
	print(reduced.samples.filtered[,-c(posName,posFilter)])
	plot.beta.densities(beta.raw[,unique(samples.filtered$sample)],"Densities of raw autosomal beta values of filtered samples")
} else {
	cat("No samples were filtered.")
}
```

### Sex mismatch

Gamete methylation can be used to check sex mismatches. First, we see the loadings of the PCA on markers in the space of samples. The second plot shows the principle components of the PCA on the samples in the space of samples.

```{r}
############## sex mismatch ##############

# Raw betas
sexprobes <- c(sexchrX,sexchrY) # chr X+Y

# EV NA omit
betaPCA <- beta.raw.sex[sexprobes,]
pcaSex <- prcomp(na.omit(betaPCA))
EV <- pcaSex$rotation
EV <- merge(EV,datsamples,by.x=0,by.y="Sentrix_Code")
colnames(EV)[match("Row.names",colnames(EV))] <- "Sentrix_Code"

# plot first males, then females: find female outlier
idx = which(EV$Gender%in%c(1)) # 1 is female
plot(EV[-idx,"PC1"],EV[-idx,"PC2"],col="orange", main=paste0("number of sex probes: ",nrow(betaPCA)," (RAW NA omitted)"),sub = "check for female outlier", xlab=paste0("EV1 (",sprintf("%1.1f",(pcaSex$sdev^2/sum(pcaSex$sdev^2))[1]*100),"%)"),ylab=paste0("EV2 (",sprintf("%1.1f",(pcaSex$sdev^2/sum(pcaSex$sdev^2))[2]*100),"%)"), xlim = c(min(EV[,"PC1"]),max(EV[,"PC1"]) + 0.25*(max(EV[,"PC1"]) - min(EV[,"PC1"])  )), ylim = c(min(EV[,"PC2"]),max(EV[,"PC2"]) ))
points(EV[idx,"PC1"],EV[idx,"PC2"],col="blue")
legend("top", pch = c(1,1), col = c("blue","orange"), legend = c("women","men"))

# then plot first females, then males: find male-outlier
plot(EV[idx,"PC1"],EV[idx,"PC2"],col="blue", main=paste0("number of sex probes: ",nrow(betaPCA)," (RAW NA omitted)"),sub = "check for male outlier", xlab=paste0("EV1 (",sprintf("%1.1f",(pcaSex$sdev^2/sum(pcaSex$sdev^2))[1]*100),"%)"),ylab=paste0("EV2 (",sprintf("%1.1f",(pcaSex$sdev^2/sum(pcaSex$sdev^2))[2]*100),"%)"), xlim = c(min(EV[,"PC1"]),max(EV[,"PC1"])+ 0.25*(max(EV[,"PC1"]) - min(EV[,"PC1"])  )), ylim = c(min(EV[,"PC2"]),max(EV[,"PC2"]) ))
points(EV[-idx,"PC1"],EV[-idx,"PC2"],col="orange")
legend("top", pch = c(1,1), col = c("blue","orange"), legend = c("women","men"))

# find potential outlier and save a list
get.outliers = function(x, SDrange = 3){
    out <- which(x>mean(x)+sd(x)*SDrange) # upper X
    out <- c(out,which(x<mean(x)-sd(x)*SDrange)) # lower X
    return(out)
}    
mark.outliers  = function(x,y, color, labels){
	out = unique(get.outliers(x),get.outliers(y))
	if (length(out)>0) {
       		text(x[out],y[out],label=labels[out],col=color,adj=c(0,0),cex=0.7)
	}
	return(labels[out])
}

sex.mismatches = mark.outliers(EV[idx,"PC1"], EV[idx,"PC2"], "blue", EV$Sentrix_Code[idx])
sex.mismatches = c(sex.mismatches, mark.outliers(EV[-idx,"PC1"], EV[-idx,"PC2"], "red",EV$Sentrix_Code[-idx]) )
```

The samples whose loadings are more than three standard deviations from the mean of the loadings with same assigned sex according to the first two principal components:

````{r}
print(sex.mismatches)
rm(sex.mismatches)
```


```{r}
resSexCheck <- data.frame(Sentrix_Code=EV$Sentrix_Code, Sex=EV$Gender, EV1=EV[,"PC1"], EV2=EV[,"PC2"])

# PC NA omit
betaPCA <- beta.raw.sex[sexprobes,]
pcaSex <- prcomp(t(na.omit(betaPCA)))
EV <- pcaSex$x
EV <- merge(EV,datsamples,by.x=0,by.y="Sentrix_Code")
colnames(EV)[match("Row.names",colnames(EV))] <- "Sentrix_Code"

# plot first males, then females: find female outlier
idx = which(EV$Gender%in%c(1)) # 1 is female
plot(EV[-idx,"PC1"],EV[-idx,"PC2"],col="orange", main=paste0("number of sex probes: ",nrow(betaPCA)," (RAW NA omitted)"),sub = "check for female outlier", xlab=paste0("PC1 (",sprintf("%1.1f",(pcaSex$sdev^2/sum(pcaSex$sdev^2))[1]*100),"%)"),ylab=paste0("PC2 (",sprintf("%1.1f",(pcaSex$sdev^2/sum(pcaSex$sdev^2))[2]*100),"%)"), xlim = c(min(EV[,"PC1"]),max(EV[,"PC1"])+ 0.25*( max(EV[,"PC1"]) - min(EV[,"PC1"])  )), ylim = c(min(EV[,"PC2"]),max(EV[,"PC2"]) ))
points(EV[idx,"PC1"],EV[idx,"PC2"],col="blue")
legend("top", pch = c(1,1), col = c("blue","orange"), legend = c("women","men"))

# then plot first females, then males: find male-outlier
plot(EV[idx,"PC1"],EV[idx,"PC2"],col="blue", main=paste0("number of sex probes: ",nrow(betaPCA)," (RAW NA omitted)"),sub = "check for male outlier", xlab=paste0("PC1 (",sprintf("%1.1f",(pcaSex$sdev^2/sum(pcaSex$sdev^2))[1]*100),"%)"),ylab=paste0("PC2 (",sprintf("%1.1f",(pcaSex$sdev^2/sum(pcaSex$sdev^2))[2]*100),"%)"), xlim = c(min(EV[,"PC1"]),max(EV[,"PC1"])+ 0.25*(max(EV[,"PC1"]) - min(EV[,"PC1"])  )), ylim = c(min(EV[,"PC2"]),max(EV[,"PC2"]) ))
points(EV[-idx,"PC1"],EV[-idx,"PC2"],col="orange")
legend("top", pch = c(1,1), col = c("blue","orange"), legend = c("women","men"))

sex.mismatches = mark.outliers(EV[idx,"PC1"], EV[idx,"PC2"], "blue", EV$Sentrix_Code[idx])
 sex.mismatches = c(sex.mismatches, mark.outliers(EV[-idx,"PC1"], EV[-idx,"PC2"], "red",EV$Sentrix_Code[-idx]) )

resSexCheck <- cbind(resSexCheck,data.frame(PC1=EV[,"PC1"],PC2=EV[,"PC2"]))
write.csv(resSexCheck,file = paste0(outputdir,"/results-PCAsex_raw","_",projectname,"-",formatted.time,".csv"),row.names=F,quote=F)

```

The samples whose values are more than three standard deviations from the mean of the samples with same assigned sex according to the first two principal components:

````{r}
print(sex.mismatches)

```

To further investigate possible sex mismatches, it might be useful to compare SNP alleles. Some of these are recorded on the methylation array for this purpose and can be obtained using  the R-package \texttt{minfi}: First read the rgSet from the concerned samples and then run the function \textbf{getSnpBeta} from minfi to get a list of rs-identifiers and corresponding SNP information. These can be compared with genotype information from other sources to clarify whether samples were mixed up. 


The previous calculations provide all information needed to filter the samples and make a tab-separated file *samplesfilefinal* for further use. 
For this analysis, `r samplesfilefinal` was used as list of samples for the final preprocessing steps. 
