## PART 1: Data and parameters

```{r, message = FALSE, echo = FALSE, warning = FALSE}
#####################

 #read data
 cwd<-getwd()

 setwd(idatdir)
#  filenames <- unique(gsub("_Grn.idat","",gsub("_Red.idat","",dir(pattern="\\.idat$",recursive=TRUE),ignore.case=T),ignore.case=T))[1:10]
 filenames <- unique(gsub("_Grn.idat","",gsub("_Red.idat","",dir(pattern="\\.idat$",recursive=TRUE),ignore.case=T),ignore.case=T))

length.of.folder = sapply(gregexpr2("/",filenames), max) 
# can be used to replace the nchar - 18 # TODO

sample_names = sapply(filenames, function(s){substring(s, first = (nchar(s) - 18))})
# dim(filenames) <- c(1,length(filenames))
# as list : sample_names = apply(filenames, c(1,2), function(s){substring(s, first = (nchar(s) - 18))})

# read samples
datsamples <- read.table(file=samplesfile,header=T,as.is=T,sep="\t", colClasses = "character")
datsamples$Gender = as.numeric(datsamples$Gender)
# provide a new column with sample identifier
if(!all(c("Sentrix_ID","Sentrix_Position")%in%colnames(datsamples))){stop("samplesfile must provide Sentrix_ID and Sentrix_Position.")}
datsamples$Sentrix_Code = paste(datsamples$Sentrix_ID,datsamples$Sentrix_Position,sep = "_")
datsamples$MFGender = ifelse(datsamples$Gender == 0, "M","F")

# read final samples
datsamplesfinal <- read.table(file=samplesfilefinal,header=T,as.is=T,sep="\t", colClasses = "character")
datsamplesfinal$Gender = as.numeric(datsamplesfinal$Gender)
# provide a new column with sample identifier
if(!all(c("Sentrix_ID","Sentrix_Position")%in%colnames(datsamplesfinal))){stop("Final samplesfile must provide Sentrix_ID and Sentrix_Position.")}
datsamplesfinal$Sentrix_Code = paste(datsamplesfinal$Sentrix_ID,datsamplesfinal$Sentrix_Position,sep = "_")
# datsamplesfinal$MFGender = ifelse(datsamplesfinal$Gender == 0, "M","F")


# filter filenames list with combined samples of samplesfile and samplesfile_final
 filenames_idx <- sapply(union(datsamplesfinal$Sentrix_Code,datsamples$Sentrix_Code), function(x){grep(x,filenames)})
  
 filenames_final_idx <- sapply(datsamplesfinal$Sentrix_Code, function(x){grep(x,filenames)})
 filenames_final_vector <- filenames[filenames_final_idx]
 
 filenames = filenames[filenames_idx] # !! this changes filenames object! 
 dim(filenames) <- c(1,length(filenames))


detectionP.local <- function (rgSet, type = "m+u")
{
    locusNames <- getManifestInfo(rgSet, "locusNames")
    detP <- matrix(NA_real_, ncol = ncol(rgSet), nrow = length(locusNames),
        dimnames = list(locusNames, sampleNames(rgSet))) 
    controlIdx <- getControlAddress(rgSet, controlType = "NEGATIVE")
    r <- getRed(rgSet)
    rBg <- r[controlIdx, ]
    rMu <- colMedians(rBg)
    rSd <- colMads(rBg)
    g <- getGreen(rgSet) 
    gBg <- g[controlIdx, ]
    gMu <- colMedians(gBg)
    gSd <- colMads(gBg)
    TypeII <- getProbeInfo(rgSet, type = "II")
    TypeI.Red <- getProbeInfo(rgSet, type = "I-Red")
    TypeI.Green <- getProbeInfo(rgSet, type = "I-Green")

    minSd <- 0.000001
    for (i in 1:ncol(rgSet)) {
        # change sd to min 0.000001
        intensity <- r[TypeI.Red$AddressA, i] + r[TypeI.Red$AddressB, i]
        detP[TypeI.Red$Name, i] <- 1 - pnorm(intensity, mean = rMu[i] * 2, sd = max(rSd[i],minSd) * 2)
        intensity <- g[TypeI.Green$AddressA, i] + g[TypeI.Green$AddressB, i] 
        detP[TypeI.Green$Name, i] <- 1 - pnorm(intensity, mean = gMu[i] * 2, sd = max(gSd[i],minSd) * 2)
        intensity <- r[TypeII$AddressA, i] + g[TypeII$AddressA, i]
        detP[TypeII$Name, i] <- 1 - pnorm(intensity, mean = rMu[i] + gMu[i], sd = max(rSd[i] + gSd[i],minSd))
    }
    detP
}

for(i in 1:nrow(filenames)){ # in code as is needless since i is only 1 and all files are read simultaneously
# for(i in 1:5){ # test version
	# print(i)
	### 1 line here different
	# RGset <- read.metharray(filenames[i,], verbose=TRUE) # formerly: read.450k()
	RGset <- read.metharray(paste0(idatdir,"/",filenames[i,]), verbose=TRUE, force=TRUE) #
	

	RGset <- bgcorrect.illumina(RGset)  # Illumina background subtraction 
	
	# add: make WBC estimation with samples from *final* samplesfile
	# integrated into for loop 18th of November 2022 FGC

		WBCok = ifelse(exists("estimateWBCs"), estimateWBCs, FALSE)
		 if(WBCok%in%TRUE){
		# if(WBCok%in%TRUE && filenames[i,] %in% filenames_final_vector) { # no this does not work as reduction to final samples; the RGset must be either reduced directly or a RGset_final mst be created as read-in

		#  	filenames <- unique(gsub("_Grn.idat","",gsub("_Red.idat","",dir(pattern="\\.idat$",recursive=TRUE),ignore.case=T),ignore.case=T))
		#  	length.of.folder = max(gregexpr2("/", filenames[1])[[1]]) # we assume the pattern is identical for all measurements
		#  	length.of.folder = sapply(gregexpr2("/",filenames), max)  # different paths 
		#  	sample_names = substring(filenames, first = length.of.folder + 1)
		#  	filenames = filenames[sample_names%in%datsamplesfinal$Sentrix_Code]
 	
 		# added 18.04.18 by FGC: minfi WBC estimation based on RGsets only 
 	
		#  	RGset <- read.metharray(filenames, verbose=TRUE, force=TRUE) # formerly: read.450k()
		#  	RGset <- bgcorrect.illumina(RGset)  # Illumina background subtraction 
 	
 		est_wbc <- estimateCellCounts(RGset, compositeCellType = "Blood", cellTypes = c("CD8T","CD4T", "NK","Bcell","Mono","Gran"),
 				 sex = datsamples$MFGender[match(sample_names,datsamples$Sentrix_Code)]) # , # )
 	                              # referencePlatform=arraytype)
 	                              
 		if(exists("est.wbc.minfi")){ # in code as is needless since i is only 1 and all files are read simultaneously
 			est.wbc.minfi = rbind(est.wbc.minfi, est_wbc)
 		} else {
 			est.wbc.minfi = as.data.frame(est_wbc)
 		}
	} # end of WBC estimation

	
	# refine and create methylation data objects 
	
    if(exists("extractSNPs")){
	if(extractSNPs%in%"TRUE"){
        	snps <- getSnpBeta(RGset)
       		 if(exists("snp.extraction")){
       		     snp.extraction <- cbind(snp.extraction, snps)
       		 } else {
       		     snp.extraction <- snps
       		 }
	}
    }
	
	# Type II probes
	TypeII.Name <- getProbeInfo(RGset, type = "II")$Name
	TypeII.Green <- getGreen(RGset)[getProbeInfo(RGset, type = "II")$AddressA,]
	TypeII.Red <- getRed(RGset)[getProbeInfo(RGset, type = "II")$AddressA,]
	rownames(TypeII.Red) <- TypeII.Name
	colnames(TypeII.Red) <- sampleNames(RGset)
	rownames(TypeII.Green) <- TypeII.Name
	colnames(TypeII.Green) <- sampleNames(RGset)

	# Type I probes, split into green and red channels
	TypeI.Green.Name <- getProbeInfo(RGset, type = "I-Green")$Name
	TypeI.Green.M <- getGreen(RGset)[getProbeInfo(RGset, type = "I-Green")$AddressB,]
	rownames(TypeI.Green.M) <- TypeI.Green.Name
	colnames(TypeI.Green.M) <- sampleNames(RGset)
	TypeI.Green.U <- getGreen(RGset)[getProbeInfo(RGset, type = "I-Green")$AddressA,]
	rownames(TypeI.Green.U) <- TypeI.Green.Name
	colnames(TypeI.Green.U) <- sampleNames(RGset)

	TypeI.Red.Name <- getProbeInfo(RGset, type = "I-Red")$Name
	TypeI.Red.M <- getRed(RGset)[getProbeInfo(RGset, type = "I-Red")$AddressB,]
	rownames(TypeI.Red.M) <- TypeI.Red.Name
	colnames(TypeI.Red.M) <- sampleNames(RGset)
	TypeI.Red.U <- getRed(RGset)[getProbeInfo(RGset, type = "I-Red")$AddressA,]
	rownames(TypeI.Red.U) <- TypeI.Red.Name
	colnames(TypeI.Red.U) <- sampleNames(RGset)

	### until next ### different: control defined 
	control=getProbeInfo(RGset, type = "Control") # needed also for lowlevel QC

	#BSC1 control probes - 450k: 16:18 ### 1:18 sind f??r 45ok die Zeilen control$Type=="BISULFITE CONVERSION I" & (control$Color %in% c("Green","LimeGreen","Lime")) 
        BSCI.Green.Name = control[control$Type=="BISULFITE CONVERSION I" & (control$Color %in% c("Green","LimeGreen","Lime")),]$ExtendedType
        BSCI.Green <- matrix(NA_real_, ncol = ncol(RGset), nrow = length(BSCI.Green.Name), dimnames = list(BSCI.Green.Name, sampleNames(RGset)))
        BSCI.Green[BSCI.Green.Name,] <- getGreen(RGset)[control[control$Type=="BISULFITE CONVERSION I" & (control$Color %in% c("Green","LimeGreen","Lime")),]$Address,]
        # 450k: 22:24 ### 22:24 entsprechen control$Type=="BISULFITE CONVERSION I" & (control$Color %in% c("Purple","Red","Tomato"))
        BSCI.Red.Name = control[control$Type=="BISULFITE CONVERSION I" & (control$Color %in% c("Purple","Red","Tomato")),]$ExtendedType
        BSCI.Red <- matrix(NA_real_, ncol = ncol(RGset), nrow = length(BSCI.Red.Name), dimnames = list(BSCI.Red.Name, sampleNames(RGset)))
        BSCI.Red[BSCI.Red.Name,] <- getRed(RGset)[control[control$Type=="BISULFITE CONVERSION I" & (control$Color %in% c("Purple","Red","Tomato")),]$Address,]

        #BSC2 control probes - 450k: 28:31 ### s.o. 
        BSCII.Red.Name = control[control$Type=="BISULFITE CONVERSION II",]$ExtendedType
        BSCII.Red <- matrix(NA_real_, ncol = ncol(RGset), nrow = length(BSCII.Red.Name), dimnames = list(BSCII.Red.Name, sampleNames(RGset)))
        BSCII.Red[BSCII.Red.Name,] <- getRed(RGset)[control[control$Type=="BISULFITE CONVERSION II",]$Address,]
	
	# select control probes green channel by ExtendedType:
        # STAINING, EXTENSION, HYBRIDISATION, TARGET REMOVAL, Specificity I+II, NON POLYMORPHIC
        extendedTypeGreen=c("Biotin (High)","Extension (C)","Extension (G)","Hyb (High)","Hyb (Medium)","Hyb (Low)","Target Removal 1","Target Removal 2","GT Mismatch 1 (PM)","GT Mismatch 2 (PM)","GT Mismatch 3 (PM)","NP (C)","NP (G)")
        extendedTypeRed=c("DNP (High)","Extension (A)","Extension (T)","GT Mismatch 4 (PM)","GT Mismatch 5 (PM)","GT Mismatch 6 (PM)","Specificity 1","Specificity 2","Specificity 3","NP (A)","NP (T)")

        controlExtGreen <- control[control$ExtendedType %in% extendedTypeGreen,]
        controlExtRed <- control[control$ExtendedType %in% extendedTypeRed,]

        ctlext.Red.Name = controlExtRed$ExtendedType
        ctlext.Red <- matrix(NA_real_, ncol = ncol(RGset), nrow = length(ctlext.Red.Name), dimnames = list(ctlext.Red.Name, sampleNames(RGset)))
        ctlext.Red[ctlext.Red.Name,] <- getRed(RGset)[controlExtRed$Address,]
        ctlext.Green.Name = controlExtGreen$ExtendedType
        ctlext.Green <- matrix(NA_real_, ncol = ncol(RGset), nrow = length(ctlext.Green.Name), dimnames = list(ctlext.Green.Name, sampleNames(RGset)))
        ctlext.Green[ctlext.Green.Name,] <- getGreen(RGset)[controlExtGreen$Address,]

	### the upper selection of information of extended type is in a more condensed form by AT ; normalization has to be added by hand 

	#Normalisation
	normC.Green.Name=control[control[,2]=='NORM_C',4]
	normC.Green <- matrix(NA_real_, ncol = ncol(RGset), nrow = length(normC.Green.Name), dimnames = list(normC.Green.Name, sampleNames(RGset)))
	normC.Green[normC.Green.Name,] <- getGreen(RGset)[control[control[,2]=='NORM_C',1],]
	normG.Green.Name=control[control[,2]=='NORM_G',4]
	normG.Green <- matrix(NA_real_, ncol = ncol(RGset), nrow = length(normG.Green.Name), dimnames = list(normG.Green.Name, sampleNames(RGset)))
	normG.Green[normG.Green.Name,] <- getGreen(RGset)[control[control[,2]=='NORM_G',1],]
	normA.Red.Name=control[control[,2]=='NORM_A',4]
	normA.Red <- matrix(NA_real_, ncol = ncol(RGset), nrow = length(normA.Red.Name), dimnames = list(normA.Red.Name, sampleNames(RGset)))
	normA.Red[normA.Red.Name,] <- getRed(RGset)[control[control[,2]=='NORM_A',1],]
	normT.Red.Name=control[control[,2]=='NORM_T',4]
	normT.Red <- matrix(NA_real_, ncol = ncol(RGset), nrow = length(normT.Red.Name), dimnames = list(normT.Red.Name, sampleNames(RGset)))
	normT.Red[normT.Red.Name,] <- getRed(RGset)[control[control[,2]=='NORM_T',1],]
	
	### here is code added for lowlevel QC (until combine ctrl probe intensities)
	
	# added for lowlevel QC
        #build dataframe with all control probes
        bufR <- intersect(rownames(getRed(RGset)),control[,1])
        bufG <- intersect(rownames(getGreen(RGset)),control[,1])
        control_ok.Red <- control[control[,1] %in% bufR,] # some controls seem to be in annotation file but not on array
        control_ok.Green <- control[control[,1] %in% bufG,] # some controls seem to be in annotation file but not on array

        # added for lowlevel QC
        # controls probes for red and green channel
        ctrl.complete.Red <- matrix(NA_real_, ncol = ncol(RGset), nrow = nrow(control_ok.Red), dimnames = list(control_ok.Red[,4], sampleNames(RGset)))
        ctrl.complete.Red <- getRed(RGset)[control_ok.Red[,1],]
        ctrl.complete.Green <- matrix(NA_real_, ncol = ncol(RGset), nrow = nrow(control_ok.Green), dimnames = list(control_ok.Green[,4], sampleNames(RGset)))
        ctrl.complete.Green <- getGreen(RGset)[control_ok.Green[,1],]
	
	#combine ctrl probe intensities ### shorter 
        ctrl = rbind(as.matrix(BSCI.Green), as.matrix(BSCI.Red), as.matrix(BSCII.Red), (ctlext.Red), (ctlext.Green),as.matrix(normC.Green),as.matrix(normG.Green), as.matrix(normA.Red),as.matrix(normT.Red))

	#detection p-values
	dp = detectionP.local(RGset, type = "m+u") ### formerly: detectionP(RGset, type = "m+u")

	
	#add data for the new samples
	if(exists("TypeII.Red.All")) {
	TypeII.Red.All <- cbind(TypeII.Red.All,TypeII.Red)
	TypeII.Green.All <- cbind(TypeII.Green.All,TypeII.Green)
	TypeI.Red.M.All <- cbind(TypeI.Red.M.All,TypeI.Red.M)
	TypeI.Red.U.All <- cbind(TypeI.Red.U.All,TypeI.Red.U)
	TypeI.Green.M.All <- cbind(TypeI.Green.M.All,TypeI.Green.M)
	TypeI.Green.U.All <- cbind(TypeI.Green.U.All,TypeI.Green.U)
	ctrl.all <- rbind(ctrl.all, t(ctrl))
	## added for lowlevel QC by AT 
	ctrl.complete.Red.all <- cbind(ctrl.complete.Red.all, ctrl.complete.Red) # added for lowlevel QC
        ctrl.complete.Green.all <- cbind(ctrl.complete.Green.all, ctrl.complete.Green) # added for lowlevel QC
	dp.all <- cbind(dp.all, dp)
	} else {
	TypeII.Red.All <- TypeII.Red                                                                                                                                  
	TypeII.Green.All <- TypeII.Green 
	TypeI.Red.M.All <- TypeI.Red.M 
	TypeI.Red.U.All <- TypeI.Red.U 
	TypeI.Green.M.All <- TypeI.Green.M 
	TypeI.Green.U.All <- TypeI.Green.U     
	ctrl.all <- t(ctrl)
	ctrl.complete.Red.all <- ctrl.complete.Red # added for lowlevel QC
        ctrl.complete.Green.all <- ctrl.complete.Green # added for lowlevel QC
        dp.all <- dp
	}
}

	

setwd(cwd)	

# save.image("test_runbyhand.Rdata")
```

### Read data and parameters

We are working with data from directory 
`r paste(strsplit(idatdir, split = "/")[[1]][-1], collapse = "/ ")` 
which contains $`r length(filenames) `$ idat files.
The annotationfile used is 
`r paste(strsplit(annotationfile, split = "/")[[1]][-1], collapse = "/ ")`
 - if problems occur with annotation, please have a look at [Illumina downloads](https://support.illumina.com/array/downloads.html)    *Infimum Methylation EPIC Product files*. 


Output is directed to  `r paste(strsplit(outputdir,split = "/")[[1]][-1], collapse = "/ ")`.
 We use samples listed in `r paste(strsplit(samplesfile, split = "/")[[1]][-1], collapse = "/ ")`\newline
  for quality control. 

As given in *parameterfile.R*, the following parameters were used: 

parameter                     | value
------------------------------|-------------------------------------
arraytype                     | `r arraytype`
detPthreshold                 | $`r detPthreshold`$
callrate.thres                | $`r callrate.thres`$
filterOutlierCtrlQC           | `r filterOutlierCtrlQC`
QuantileNormalize             | `r QuantileNormalize`
InterQuartileRangeCalculation | `r InterQuartileRangeCalculation`
estimateWBCs                  | `r ifelse(exists("estimateWBCs"),estimateWBCs, "not provided, assumed FALSE")`
extractSNPs                   |  `r ifelse(exists("extractSNPs"),extractSNPs, "not provided, assumed FALSE")`

Further we interpret the values for the gender in the samplesfile as **1=female** and **0=male**. 

```{r}
if(QuantileNormalize == TRUE && length(samplesfilefinal)==0){
	cat("In order to run Part 3: Quantile normalization and IQ-filter a final samplefile is needed.\n The samplesfile of the first and second part will be used instead.")
	samplesfilefinal <- samplesfile
}
path_depth<-length(strsplit(samplesfilefinal, split = "/")[[1]])
if(QuantileNormalize == TRUE){cat(paste0("The samples listed in\n", paste(strsplit(samplesfilefinal, split = "/")[[1]][2:(path_depth-3)], collapse = "/"),"/\n",paste0(strsplit(samplesfilefinal, split = "/")[[1]][(path_depth-2):path_depth], collapse = "/"),"\n are used for quantile normalization and calculation of outliers regarding inter-quartile-range."))}
```

When reading the data using the minfi-package we apply Illumina Background correction.
Within this process we also 

* extract control-probe information. 
* calculate detection p-values.
* estimate the white blood cell distribution assuming whole blood samples using minfi.
* separate the data by channel (red / green) and Infinium I / II type.

We use this detection p-values and control probe information for high-level quality control and the white blood cell estimations for further processing as phenodata.

### White Blood Cell estimation

If the switch estimateWBCs is set to TRUE in the parameterfile, white blood cell distributions are estimated assuming measurements from whole blood.
The estimation of White Blood Cells results in a data.frame est.wbc.minfi for further use as part of the phenodata.\newline
*Notice*: Dependencies of the results on the sample selection is possible. To avoid that the WBC estimation of fine runs is disturbed by problematic measurements, the estimation is based on samples from the final sample file only.

```{r} 
if(exists("est.wbc.minfi")){
     head(est.wbc.minfi)
}

```

### SNP data extraction

If the switch extractSNPs is set to TRUE in the parameterfile, the SNP information from the RGset is extracted and exported to the file `r gsub("/","/ ", paste0(outputdir,"/extracted-snps","_",projectname,"-",formatted.time,".csv"))`. 
This information can be used to detect mismatches in labeling of samples and often explains most of the sex mismatches.
Homozygotes should have values around $0$ and $1$, whereas heterozygotes have values close to $0.5$. 
Please consider this when comparing to the genotyping from other sources which may be coded $0$ - $1$ - $2$. 

```{r} 
if(exists("extractSNPs") && extractSNPs%in%"TRUE"){
    if(exists("snp.extraction")){
        write.csv(snp.extraction, file = paste0(outputdir,"/extracted-snps","_",projectname,"-",formatted.time,".csv"),row.names=T)
    } else {
        cat("\nAn error occurres during SNP extraction. No results are exported.\n")
    }
}
# rm(RGset, snp.extraction)
```


### Data preparation

The probes are divided by chromosome type: autosomal probes, chromosome X probes and chromosome Y probes.
For this step we need the annotationfile.

```{r}
anno=read.csv(annotationfile,as.is=TRUE, skip = 7) 
 anno=anno[,c('Infinium_Design_Type','Color_Channel', 'CHR', 'MAPINFO', 'Name')]
# get autosomal probes
 cas=anno[substr(anno$Name, 1,3)=='ch.' & !(anno$CHR %in% c('X','Y')),]
 cgs=anno[substr(anno$Name, 1,2)=='cg'& !(anno$CHR %in% c('X','Y')),]
 auto = c(cgs$Name, cas$Name)
 auto=as.matrix(auto)

# get chrX probes
 xcas=anno[substr(anno$Name, 1,3)=='ch.' & (anno$CHR %in% c('X')),]
 xcgs=anno[substr(anno$Name, 1,2)=='cg'& (anno$CHR %in% c('X')),]
 sexchrX = c(xcgs$Name, xcas$Name)
 sexchrX=as.matrix(sexchrX)

 # get chrY probes
 ycas=anno[substr(anno$Name, 1,3)=='ch.' & (anno$CHR %in% c('Y')),]
 ycgs=anno[substr(anno$Name, 1,2)=='cg'& (anno$CHR %in% c('Y')),]
 sexchrY = c(ycgs$Name, ycas$Name)
 sexchrY=as.matrix(sexchrY)
```

### High-level quality control


High level quality control includes detection p-value filter, restriction the the samples listed in the samplesfile and call rate filtering.

We calculate raw beta values for both autosomal and sex chromosome data.
In our further calculations beta value data always is processed separately for autosomes and gametes. 

* Detection p-values are illustrated in the following plots. Low p-values indicate that the signal is unlikely to be background noise. 

```{r}
# test
dpvec <- dp.all
dim(dpvec) <- c(nrow(dp.all)*ncol(dp.all),1)

# plot detection P
dpvecplot <- dpvec
dpvecplot[which(dpvecplot==0)] <- 1E-20
if(sum(dpvec>0)>0){
	hist(-log10(dpvec[!dpvec%in%0]),breaks=30, main = "detection p-values: only positive values")
}
```



```{r}
hist(-log10(dpvecplot),breaks=30, main = "detection p-values: zero values set to 1E-20")
rm(dpvecplot)
```
\pagebreak


The following table summarizes how many detection p-values are smaller than the threshold $`r detPthreshold `$ given in the *parameterfile*, or 0.01: 

 threshold           | count                                       | proportion
---------------------|---------------------------------------------|-----------------------------------------------------
$`r detPthreshold `$ | $`r length(which(dpvec<detPthreshold))`$    | $`r length(which(dpvec<detPthreshold))/length(dpvec)`$
0.01                 | $`r length(which(dpvec<0.01))`$             | $`r length(which(dpvec<0.01))/length(dpvec)`$

$`r sum(is.na(dpvec))`$ detection p-values are missing.

```{r}
#set detection p-value threshold
thres=detPthreshold
d=dp.all[rownames(TypeII.Green.All),colnames(TypeII.Green.All)]
TypeII.Green.All.d = ifelse(d<thres,TypeII.Green.All,NA)
TypeII.Red.All.d = ifelse(d<thres,TypeII.Red.All,NA)
d=dp.all[rownames(TypeI.Green.M.All),colnames(TypeI.Green.M.All)]
TypeI.Green.M.All.d = ifelse(d<thres,TypeI.Green.M.All,NA)
TypeI.Green.U.All.d = ifelse(d<thres,TypeI.Green.U.All,NA)
d=dp.all[rownames(TypeI.Red.M.All),colnames(TypeI.Red.M.All)]
TypeI.Red.M.All.d = ifelse(d<thres,TypeI.Red.M.All,NA)
TypeI.Red.U.All.d = ifelse(d<thres,TypeI.Red.U.All,NA)
#rm(dp.all,d)

samples=colnames(TypeI.Red.M.All)
```

$`r length(which(dpvec>detPthreshold))`$   ($`r round(length(which(dpvec>detPthreshold))/length(dpvec), digits = 3)*100`$ \%)   measurements are excluded because their detection p-value is bigger than $`r detPthreshold`$. Only values with a detection p-value strictly smaller the threshold are kept.
To skip this filtering, set the parameter *detPthreshold* to a value **strictly** bigger than 1 in the *parameterfile.R*.


```{r}

#autosomes ------------------------------------------------------------------
 category=auto
 markers=as.matrix(intersect(rownames(TypeII.Green.All.d), category))
# head(markers)
TypeII.Green = TypeII.Green.All.d[markers,samples]
TypeII.Red = TypeII.Red.All.d[markers,samples]
markers=intersect(rownames(TypeI.Green.M.All.d), category)
TypeI.Green.M = TypeI.Green.M.All.d[markers,samples]
TypeI.Green.U = TypeI.Green.U.All.d[markers,samples]
markers=intersect(rownames(TypeI.Red.M.All.d), category)
TypeI.Red.M = TypeI.Red.M.All.d[markers,samples]
TypeI.Red.U = TypeI.Red.U.All.d[markers,samples]

#raw betas
TypeII.betas = TypeII.Green/(TypeII.Red+TypeII.Green+100)
TypeI.Green.betas = TypeI.Green.M/(TypeI.Green.M+TypeI.Green.U+100)
TypeI.Red.betas = TypeI.Red.M/(TypeI.Red.M+TypeI.Red.U+100)
beta.raw = as.matrix(rbind(TypeII.betas,TypeI.Green.betas,TypeI.Red.betas))

# sex chromosomes ------------------------------------------------------------------

category=c(sexchrX,sexchrY)
markers=as.matrix(intersect(rownames(TypeII.Green.All.d), category))
TypeII.Green = TypeII.Green.All.d[markers,samples]
TypeII.Red = TypeII.Red.All.d[markers,samples]
markers=intersect(rownames(TypeI.Green.M.All.d), category)
TypeI.Green.M = TypeI.Green.M.All.d[markers,samples]
TypeI.Green.U = TypeI.Green.U.All.d[markers,samples]
markers=intersect(rownames(TypeI.Red.M.All.d), category)
TypeI.Red.M = TypeI.Red.M.All.d[markers,samples]
TypeI.Red.U = TypeI.Red.U.All.d[markers,samples]

#raw betas
TypeII.betas = TypeII.Green/(TypeII.Red+TypeII.Green+100)
TypeI.Green.betas = TypeI.Green.M/(TypeI.Green.M+TypeI.Green.U+100)
TypeI.Red.betas = TypeI.Red.M/(TypeI.Red.M+TypeI.Red.U+100)
beta.raw.sex = as.matrix(rbind(TypeII.betas,TypeI.Green.betas,TypeI.Red.betas))
 
```


```{r}
# calculate sample and marker calls without filter -- used for sample filter 
sample.call=colSums(!is.na(beta.raw))/nrow(beta.raw)
marker.call=rowSums(!is.na(beta.raw))/ncol(beta.raw)

# remove samples not in samplesfile
samplesinclude <- intersect(names(sample.call),datsamples$Sentrix_Code)
sample.call <- sample.call[samplesinclude]
ctrl.complete.Green.all <- ctrl.complete.Green.all[,samplesinclude]
ctrl.complete.Red.all <- ctrl.complete.Red.all[,samplesinclude]
beta.raw.sex <- beta.raw.sex[,samplesinclude]
beta.raw <- beta.raw[,samplesinclude]
```

* Identified by the samplefile, $`r length(samplesinclude)`$ samples are included in the analysis.


beta values:   | autosomes           |  sex chromosomes 
---------------|---------------------|----------------------------
dimension:     | $`r dim(beta.raw)`$ | $`r dim(beta.raw.sex)`$

````{r}
if(length(samplesinclude)<101){
	cat("\nThe sample names of the included samples:\n")
	datsamples$Sample_Name[datsamples$Sentrix_Code %in%samplesinclude]
}

plot.beta.densities = function(beta, title){
  if(!is.null(dim(beta))){
  	densities = apply(beta,2,function(x){ density(x, na.rm = TRUE) } )
  	xmax = max(sapply(densities, function(d){max(d$x)}))
  	xmin = min(sapply(densities, function(d){min(d$x)}))
  	ymax = max(sapply(densities, function(d){max(d$y)}))
  
  	plot(NA, xlim = c(xmin,xmax), ylim = c(0,ymax), main = title, ylab = "")
  	colors = rainbow(10)
  	for(i in 1:ncol(beta)){
  		lines(densities[[i]], col = colors[i%%10 + 1])
  	}
  } else if(length(beta) > 1){
    plot(density(beta, na.rm = TRUE), main = title)
  }
}
plot.beta.densities(beta.raw,"Densities of raw autosomal beta values per sample")
```



```{r}
# call-rate filtering

samples.filtered <- data.frame()
bufout <- names(sample.call[sample.call<=callrate.thres])
if (length(bufout)>0) samples.filtered <- data.frame(sample=bufout,x=NA,y=NA,filter=paste0("callrate ",callrate.thres),stringsAsFactors=F)
```
  
* There is call-rate filtering with threshold $`r callrate.thres`$.

* $`r length(bufout)`$ samples were tagged for exclusion because the call-rate was below the threshold $`r callrate.thres`$.


```{r}
# plot sample call rates
hist(sample.call,breaks=100, main = "sample call rates")
```

```{r}
# plot sample call rates
hist(sample.call,breaks=100, main = "sample call rates zoomed x-axis", xlim = c(0.85,1), sub = "zoomed x-axis is fixed to 0.85 to 1")
```

```{r}
# plot sample call rates
hist(sample.call,breaks=100, main = "sample call rates zoomed y-axis", ylim = c(0,30), sub = "zoomed y axis is fixed to 0 to 30")
```

$`r max(0,length(sample.call[which(sample.call<=0.98)]))`$ of all sample call rates are lower than 0.98,  
 and $`r length(sample.call[which(sample.call<=callrate.thres)])`$ are lower than the threshold $`r callrate.thres`$.


We have a look at the marker call rates as well:


```{r}
hist(marker.call,breaks=100, main = "marker call rates")
```


```{r}
hist(marker.call,breaks=100, main = "marker call rates zoomed y-axis", ylim = c(0,500), sub = "zoomed y axis is fixed to 0 to 500")
```  

$`r max(0,length(marker.call[which(marker.call<=0.98)]))`$ of all marker call rates are lower than 0.98,  
 and $`r length(marker.call[which(marker.call<=0.95)])`$ are lower than the threshold 0.95.


The results of the sample call rate filter are included in the export file **samples_filtered.csv** which also documents the following control-probe based quality control.


